web/src/main/java/org/cbioportal/web/StudyViewController.java

```diff
@@ -223,6 +223,43 @@
         return new ResponseEntity<>(result, HttpStatus.OK);
     }
 
+    @PreAuthorize("hasPermission(#involvedCancerStudies, 'Collection<CancerStudyId>', 'read')")
+    @RequestMapping(value = "/fusion-genes/fetch", method = RequestMethod.POST,
+        consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
+    @ApiOperation("Fetch fusion genes by study view filter")
+    public ResponseEntity<List<MutationCountByGene>> fetchFusionGenes(
+        @ApiParam(required = true, value = "Study view filter")
+        @Valid @RequestBody(required = false) StudyViewFilter studyViewFilter,
+        @ApiIgnore // prevent reference to this attribute in the swagger-ui interface
+        @RequestAttribute(required = false, value = "involvedCancerStudies") Collection<String> involvedCancerStudies,
+        @ApiIgnore // prevent reference to this attribute in the swagger-ui interface. this attribute is needed for the @PreAuthorize tag above.
+        @Valid @RequestAttribute(required = false, value = "interceptedStudyViewFilter") StudyViewFilter interceptedStudyViewFilter) throws StudyNotFoundException {
+        List<SampleIdentifier> filteredSampleIdentifiers = studyViewFilterApplier.apply(interceptedStudyViewFilter);
+        List<MutationCountByGene> result = new ArrayList<>();
+        if (!filteredSampleIdentifiers.isEmpty()) {
+            List<String> studyIds = new ArrayList<>();
+            List<String> sampleIds = new ArrayList<>();
+            studyViewFilterUtil.extractStudyAndSampleIds(filteredSampleIdentifiers, studyIds, sampleIds);
```

Comment:
The logic below seems to go beyond simple argument checking and service selection. I think this qualifies as business logic which should be encapsulated and tested in the service layer. I see that a util class is employed here ... but even still, I think this qualifies as business logic (data transformations). I would not insist on relocation, but I guess I strongly recommend keeping data transformations out of the controller handler.

Reply:
total agree and very good points! It is not just here but the whole class should be refactored to pull those data transformations out and put into respective service layer. How about we have a follow up PR instead do correction here. What do you think @sheridancbio, @zhx828?

Reply:
will keep this principle in mind when switching to the new SV table and reimplementing fetch-fusion-gene endpoint.